### PR TITLE
Add ui.explore settings to control view of explore pages (2)

### LINKED ISSUE #2908: Hide users from public view
- Gitea version (or commit ref): 1.2.3
- Git version: 2.15
- Operating system: CentOS
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
I think there should exist a config option to prevent the viewing of user accounts from public view. I have user accounts on my Gitea instance that I would not like others to see. This might be doable with templates but I am not sure at the moment.

### PR DESCRIPTION
**Pull request by zeripath**
This is an alternative PR to #13687.

Add `[ui.explore]` settings to allow restricting the
explore pages to logged in users only and to disable the users explore page.

The two proposed settings are:

- `REQUIRE_SIGNIN_VIEW`: Only allows access to the explore pages if the
user is signed in. Also restricts
  - `/api/v1/user/search`
  - `/api/v1/users/{username}`
  - `/api/v1/users/{username}/repos`
  - but does not restrict `/api/v1/users/{username}/heatmap`
- `DISABLE_USERS_PAGE`: Disables the /explore/users page

Fix #2908

Close #13687 

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: docs/content/doc/advanced/config-cheat-sheet.zh-cn.md [hunk 1]
```diff
@@ -50,6 +50,11 @@ menu:
 - `NOTICE_PAGING_NUM`: 系统提示页面每页显示的提示数量。
 - `ORG_PAGING_NUM`: 组织管理页面每页显示的组织数量。
 
+### UI - Expore (`ui.explore`)
+
+- `REQUIRE_SIGNIN_VIEW`: **false**: 仅允许已登录的用户查看探索页面。
+- `DISABLE_USERS_PAGE`: **false**: 。
```
Comment (lunny) [inline docs/content/doc/advanced/config-cheat-sheet.zh-cn.md]: 不显示用户探索页面。

**Comment by lafriks** [comment]
Is there a use case where `REQUIRE_SIGNIN_VIEW` is needed? Imho that should be always true and we should not return list of users to unauthorized users

**Comment by techknowlogick** [comment]
@lafriks `REQUIRE_SIGNIN_VIEW` is for more than just users, but all explore pages (such as repos & orgs too).

**Comment by lafriks** [comment]
imho we should remove users page completely, I don't see much use for it to be honest

**Comment by lunny** [comment]
> imho we should remove users page completely, I don't see much use for it to be honest

Maybe for a public big site, users want to find some people they want to contact.

**Comment by lunny** [comment]
Please resolve the conflicts

**Comment by zeripath** [comment]
Conflicts fixed.

File: docs/content/doc/advanced/config-cheat-sheet.zh-cn.md (hunk 1 above)
Comment (zeripath) [reply docs/content/doc/advanced/config-cheat-sheet.zh-cn.md]:
```suggestion
- `DISABLE_USERS_PAGE`: **false**: 不显示用户探索页面。
```

File: modules/setting/setting.go [hunk 2]
```diff
@@ -265,6 +270,10 @@ var (
 			Description: "Gitea (Git with a cup of tea) is a painless self-hosted Git service written in Go",
 			Keywords:    "go,git,self-hosted,gitea",
 		},
+		Explore: struct {
+			RequireSigninView bool `ini:"REQUIRE_SIGNIN_VIEW"`
+			DisableUsersPage  bool `ini:"DISABLE_USERS_PAGE"`
```
Comment (lunny) [inline modules/setting/setting.go]: I think maybe move this to service is better? We need a feature to disable all the user explore include APIs. A standalone setting for UI search seems not enough.

**Comment by art-ist** [comment]
You are checking for the `ui.explore` setting when http://gitea/api/v1/users/search is called. But 

```sh
curl -X GET "http://gitea/api/v1/users/joe" -H  "accept: application/json"
```
was our next first guess. So checking on 
```go
m.Group("/{username}", func() {
```
is required as well.

_Originally posted by @art-ist in https://github.com/go-gitea/gitea/issues/2908#issuecomment-791538960_

**Comment by zeripath** [comment]
@art-ist I wrote this PR 3 months ago in response so I cannot fully remember all of the details but...

I am absolutely certain that you cannot just block /api/v1/users/{name} and in any case it is pointless as you can just go to /{name}

File: modules/setting/setting.go (hunk 2 above)
Comment (zeripath) [reply modules/setting/setting.go]: I really think we have way too much stuff in `[service]` already - so if it's going in there it should go in `[service.explore]` I think.

Comment (noerw) [reply modules/setting/setting.go]: +1 on `[service.explore]`

**Review by noerw — APPROVED** [review]
~~Oher than the config location this LGBTTOPRTM (looks good and better than the other PR to me).~~

~~I think this more general config will serve more users concerned about privacy, a more finegrained control like `ONLY_SHOW_USERS_WITH_PUBLIC_REPOS` could be added later as needed~~

I thought more about this, see https://github.com/go-gitea/gitea/issues/2908#issuecomment-796106581 . In short:
We should probably allow for more flexibility and don't add a bool flag `DISABLE_USERS_PAGE`, but `USERS_PAGE` with 3 states `OPT_OUT`, `OPT_IN`, `DISABLED`, (where the first is the current behaviour, and the second can be implemented later)

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
🚀

**Review by noerw — APPROVED** [review]
Thinking about it again again, the opt-in feature can also be realised via a separate flag, so please ignore my above comment

**Pull request merged**